lib/src/clojure_lsp/refactor/edit.clj

```diff
@@ -42,6 +42,42 @@
   [zloc p?]
   (z-filter zloc z/next p?))
 
+(defn ^:private liberal-in-range? [loc pos]
+  (if-let [node-pos (some-> loc z/node meta)]
+    (in-range? node-pos pos)
+    (let [s (z/string loc)]
+      (or (string/ends-with? s "/")
+          (string/ends-with? s ":")))))
+
+(defn right-or-up [zloc]
+  (loop [p zloc]
+    (or
+      (z/right* p)
+      (some-> p z/up* recur))))
+
+(defn by-heritability [start-zloc inherits?]
+  (loop [zloc (cond-> start-zloc
+                (= :forms (z/tag start-zloc)) z/down*)]
+    (if (z/end? zloc)
+      zloc
+      (if (inherits? zloc)
+        (if-let [inner (some-> zloc z/down* (z/find z/right* inherits?))]
+          (recur inner)
+          zloc)
+        (recur (right-or-up zloc))))))
+
+(defn find-last-by-pos-heritability [start-zloc pos]
+  (let [inherits? #(liberal-in-range? % pos)]
+    (by-heritability start-zloc inherits?)))
+
+(defn find-last-by-pos-next [zloc f pos]
+  (let [forms (z-filter zloc f #(liberal-in-range? % pos))
+        disconsider-reader-macro? (and (some #(= "?" (z/string %)) forms)
+                                       (> (count forms) 1))]
+    (last
+      (cond->> forms
+        disconsider-reader-macro? (remove (comp #(= "?" %) z/string))))))
+
 ;; From rewrite-cljs
```

Comment:
this function was included before rewrite-cljs was merged into rewrite-clj, maybe we should check if there is already a function that does this on rewrite-clj itself

Reply:
Otherwise, maybe we should a test for that function even knowing other tests would fail, a test testing the corner cases here seems to make sense, WDYT?

Reply:
rewrite-clj has `find-last-by-pos`, which perhaps it got in the merge with rewrite-cljs. But that function uses the same depth-first approach that we're trying to avoid here. I looked around rewrite-clj pretty carefully to see if there was another faster algorithm, but didn't find anything.

Reply:
> testing the corner cases here

I'm sorry, I don't follow. What should be tested?

Reply:
no problem, so there is no function that does what we need there, so my point is even if we know there are other unit tests that end up using this function, maybe we should have tests for this function especially too?

Reply:
Ah, yes, I understand. And it's a good idea!

I've added several tests for `edit/find-at-pos` that confirm that it navigates to the correct node, accounting for as many edge cases as I can think of.

`edit/find-at-pos` was originally called `edit/find-last-by-pos`. That function had special handling for two corner cases: incomplete nodes and the `#?` reader macro.

First let's talk about the reader macro, i.e. code like `#?(:clj (+ 1 2))` or `#?@(:clj (+ 1 2))`. I don't think we need special handling for it anymore. It looks like the special handling was added as part of fixing #332, which was about completion in files with reader macros. The tests that were added as part of #332 still pass, even without any special handling. I suspect that the special handling was no longer necessary as soon as rewrite-clj learned to parse reader macro nodes, around v1, which was soon after the fix for #332 was applied. So, I think we're safe to remove the special handling. Does that sound right @ericdallo?

I've extended the tests around incomplete nodes, code like `foo/`. Those tests exposed some problems with how my changes handled things. The fix was to ensure that all nodes, even ones that initially appear incomplete, have position metadata. (See `parser/replace-preserving-meta`.)  That moves all the special handling out of `edit` and into `parser`, so I added tests around `parser/safe-zloc-of-string` and `parser/to-loc`. In the process, I also fixed a long-standing bug where nodes that followed an incomplete node would have incorrect column numbers. (See `parser/zero-width-space` and the way that it is used to keep the column numbers correct.)

Reply:
> Does that sound right @ericdallo?

Yes! Since the tests I added are still passing looks good to me